File: packages/backend/sdk-python/src/fluxis/models/webhook.py
"""Webhook models."""

from __future__ import annotations

from dataclasses import dataclass
from typing import cast

from fluxis.models.common import (
    Paginated,
    WebhookEventType,
    _as_dict,
    _optional_bool,
    _optional_str,
    parse_list,
)


@dataclass
class WebhookCreateRequest:
    url: str
    event_type: WebhookEventType
    description: str | None = None


@dataclass
class WebhookUpdateUrlRequest:
    url: str


@dataclass
class Webhook:
    id: str
    url: str
    event_type: WebhookEventType
    enabled: bool
    account_id: str | None = None
    description: str | None = None
    secret: str | None = None
    created_at: str | None = None
    updated_at: str | None = None

    @classmethod
    def from_dict(cls, data: object) -> Webhook:
        payload = _as_dict(data)
        webhook_id = _optional_str(payload, "id")
        url = _optional_str(payload, "url")
        event_type = _optional_str(payload, "event_type")
        enabled = _optional_bool(payload, "enabled")
        if webhook_id is None or url is None or event_type is None or enabled is None:
            raise ValueError("Webhook response missing required fields")
        return cls(
            id=webhook_id,
            url=url,
            event_type=cast("WebhookEventType", event_type),
            enabled=enabled,
            account_id=_optional_str(payload, "account_id"),
            description=_optional_str(payload, "description"),
            secret=_optional_str(payload, "secret"),
            created_at=_optional_str(payload, "created_at"),
            updated_at=_optional_str(payload, "updated_at"),
        )


@dataclass
class WebhookLog:
    id: str
    webhook_id: str
    event_type: WebhookEventType
    account_id: str | None = None
    response_status: int | None = None
    response_body: str | None = None
    duration_ms: int | None = None
    error: str | None = None
    created_at: str | None = None

    @classmethod
    def from_dict(cls, data: object) -> WebhookLog:
        payload = _as_dict(data)
        log_id = _optional_str(payload, "id")
        webhook_id = _optional_str(payload, "webhook_id")
        event_type = _optional_str(payload, "event_type")
        if log_id is None or webhook_id is None or event_type is None:
            raise ValueError("Webhook log response missing required fields")
        response_status_raw = payload.get("response_status")
        response_status = int(response_status_raw) if isinstance(response_status_raw, int) else None
        duration_raw = payload.get("duration_ms")
        duration_ms = int(duration_raw) if isinstance(duration_raw, int) else None
        return cls(
            id=log_id,
            webhook_id=webhook_id,
            event_type=cast("WebhookEventType", event_type),
            account_id=_optional_str(payload, "account_id"),
            response_status=response_status,
            response_body=_optional_str(payload, "response_body"),
            duration_ms=duration_ms,
            error=_optional_str(payload, "error"),
            created_at=_optional_str(payload, "created_at"),
        )


@dataclass
class ListWebhookLogsOptions:
    page: int | None = None
    page_size: int | None = None


def parse_webhook_list(data: object) -> list[Webhook]:
    return parse_list(data, Webhook)


def parse_webhook_log_page(data: object) -> Paginated:
    return Paginated.from_dict(data, WebhookLog)

File: packages/backend/sdk-csharp/src/Fluxis/Models/Webhook.cs
using System.Text.Json.Serialization;

namespace Fluxis.Models;

/// <summary>Webhook event type values.</summary>
public static class WebhookEventTypes
{
    /// <summary>Payment request status updates.</summary>
    public const string PaymentRequest = "payment_request";
    /// <summary>Incoming transfer notifications.</summary>
    public const string IncomingTransfer = "incoming_transfer";
    /// <summary>Refund notifications.</summary>
    public const string Refund = "refund";
}

/// <summary>
/// Paginated API response wrapper.
/// </summary>
/// <typeparam name="T">Item type in the data array.</typeparam>
public sealed class PaginatedResponse<T>
{
    /// <summary>Page of results.</summary>
    [JsonPropertyName("data")]
    public List<T> Data { get; set; } = new();

    /// <summary>Current page number (1-based).</summary>
    [JsonPropertyName("page")]
    public int Page { get; set; }

    /// <summary>Number of items per page.</summary>
    [JsonPropertyName("page_size")]
    public int PageSize { get; set; }

    /// <summary>Total number of matching items.</summary>
    [JsonPropertyName("total")]
    public int Total { get; set; }

    /// <summary>Total number of pages.</summary>
    [JsonPropertyName("total_pages")]
    public int TotalPages { get; set; }
}

/// <summary>
/// Request body for creating an account-scoped webhook.
/// </summary>
public sealed class WebhookCreateRequest
{
    /// <summary>Webhook URL to receive events.</summary>
    [JsonPropertyName("url")]
    public string Url { get; set; } = string.Empty;

    /// <summary>Event type to subscribe to.</summary>
    [JsonPropertyName("event_type")]
    public string EventType { get; set; } = string.Empty;

    /// <summary>Optional description.</summary>
    [JsonPropertyName("description")]
    public string? Description { get; set; }
}

/// <summary>
/// Request body for updating a webhook URL.
/// </summary>
public sealed class WebhookUpdateUrlRequest
{
    /// <summary>New webhook URL.</summary>
    [JsonPropertyName("url")]
    public string Url { get; set; } = string.Empty;
}

/// <summary>
/// Webhook entity returned by the API.
/// </summary>
public sealed class Webhook
{
    /// <summary>Webhook ID.</summary>
    [JsonPropertyName("id")]
    public string Id { get; set; } = string.Empty;

    /// <summary>Account ID this webhook belongs to.</summary>
    [JsonPropertyName("account_id")]
    public string? AccountId { get; set; }

    /// <summary>Webhook URL.</summary>
    [JsonPropertyName("url")]
    public string Url { get; set; } = string.Empty;

    /// <summary>Subscribed event type.</summary>
    [JsonPropertyName("event_type")]
    public string? EventType { get; set; }

    /// <summary>Whether the webhook is enabled.</summary>
    [JsonPropertyName("enabled")]
    public bool Enabled { get; set; }

    /// <summary>Optional description.</summary>
    [JsonPropertyName("description")]
    public string? Description { get; set; }

    /// <summary>HMAC secret for signature verification (returned on create).</summary>
    [JsonPropertyName("secret")]
    public string? Secret { get; set; }

    /// <summary>Creation timestamp.</summary>
    [JsonPropertyName("created_at")]
    public string? CreatedAt { get; set; }

    /// <summary>Last update timestamp.</summary>
    [JsonPropertyName("updated_at")]
    public string? UpdatedAt { get; set; }
}

/// <summary>
/// Webhook delivery log entry.
/// </summary>
public sealed class WebhookLog
{
    /// <summary>Log entry ID.</summary>
    [JsonPropertyName("id")]
    public string Id { get; set; } = string.Empty;

    /// <summary>Account ID.</summary>
    [JsonPropertyName("account_id")]
    public string? AccountId { get; set; }

    /// <summary>Webhook ID.</summary>
    [JsonPropertyName("webhook_id")]
    public string? WebhookId { get; set; }

    /// <summary>Event type that was delivered.</summary>
    [JsonPropertyName("event_type")]
    public string? EventType { get; set; }

    /// <summary>HTTP response status from the webhook endpoint.</summary>
    [JsonPropertyName("response_status")]
    public int? ResponseStatus { get; set; }

    /// <summary>Response body from the webhook endpoint.</summary>
    [JsonPropertyName("response_body")]
    public string? ResponseBody { get; set; }

    /// <summary>Delivery duration in milliseconds.</summary>
    [JsonPropertyName("duration_ms")]
    public int? DurationMs { get; set; }

    /// <summary>Error message if delivery failed.</summary>
    [JsonPropertyName("error")]
    public string? Error { get; set; }

    /// <summary>Raw payload bytes.</summary>
    [JsonPropertyName("payload")]
    public List<int>? Payload { get; set; }

    /// <summary>Creation timestamp.</summary>
    [JsonPropertyName("created_at")]
    public string? CreatedAt { get; set; }
}
